### PR TITLE
feat: Add feature view versioning support to FAISS online store

### LINKED ISSUE #6173: Add feature view versioning support to Faiss online store
## Summary

Add versioned feature view read/write support to the `Faiss` online store, so that version-qualified feature references (e.g., `driver_stats@v2:trips_today`) resolve to the correct versioned data.

## Context

Feature view versioning was introduced in #6101. Currently, only the SQLite online store supports versioned reads/writes. All other online stores raise `VersionedOnlineReadNotSupported` when a versioned ref is used.

The implementation pattern (see `sqlite.py` for reference) involves:
1. **Write path**: When `enable_online_feature_view_versioning` is enabled, materialize data into a version-specific table/key namespace (e.g., `{project}_{fv_name}_v{version}`)
2. **Read path**: Route version-qualified lookups to the correct versioned table/key namespace

## Reference implementation

- SQLite online store: `sdk/python/feast/infra/online_stores/sqlite.py`
- Store file: `sdk/python/feast/infra/online_stores/faiss_online_store.py`

## Related

Part of #2728

### PR DESCRIPTION
**Pull request by YassinNouh21**
## Summary
- When `enable_online_feature_view_versioning` is enabled, FAISS indices are namespaced by versioned table keys (e.g. `project_driver_stats_v2`) so multiple feature view versions coexist in memory
- Refactored singleton `_index`/`_in_memory_store` to per-table dictionaries (`_indices`/`_in_memory_stores`)
- Reuses shared `compute_table_id()` from `helpers.py` for consistency with PostgreSQL and MySQL stores (#6193)
- Registered `FaissOnlineStore` in `_check_versioned_read_support()` gate

## What this PR does / why we need it
Closes #6173

Feature view versioning was introduced in #6101, but version-qualified feature references (e.g. `driver_stats@v2:trips_today`) only worked with SQLite, PostgreSQL, and MySQL online stores. All other online stores raised `VersionedOnlineReadNotSupported` when a versioned ref was used.

This PR adds versioned read/write support to the FAISS online store, following the same pattern as the PostgreSQL/MySQL implementation (#6193):

- **Write path**: when `enable_online_feature_view_versioning` is enabled, `_table_id` appends a `_v{N}` suffix to the in-memory key namespace, routing materialization to the correct versioned index
- **Read path**: version-qualified lookups (`@v2`) set `projection.version_tag` on the feature view; `_table_id` picks this up and routes to the correct index. `projection.version_tag` takes priority over `current_version_number`
- **Multi-table support**: replaced single class-level `_index`/`_in_memory_store` with per-table `_indices`/`_in_memory_stores` dicts
- **Gate**: `_check_versioned_read_support` in `online_store.py` now allows `FaissOnlineStore` through

## Which issue(s) this PR fixes
Part of #2728

## Checks
- [x] I've made sure the tests are passing.
- [x] My commits are signed off (`git commit -s`)
- [x] My PR title follows [conventional commits](https://www.conventionalcommits.org/) format

## Testing Strategy
- [x] Unit tests
- [ ] Integration tests
- [ ] Manual tests
- [ ] Testing is not required for this change

Added 16 unit tests covering `_table_id` with versioning disabled/enabled, `projection.version_tag` priority over `current_version_number`, version 0 edge case, versioned read support gate, versioned write/read round trips, namespace isolation between versions, missing index handling, and teardown cleanup.

---

<a href="https://app.devin.ai/review/feast-dev/feast/pull/6256" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: sdk/python/feast/errors.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ class VersionedOnlineReadNotSupported(FeastError):
     def __init__(self, store_name: str, version: int):
         super().__init__(
             f"Versioned feature reads (@v{version}) are not yet supported by {store_name}. "
-            f"Currently only SQLite, PostgreSQL, and MySQL support version-qualified feature references. "
+            f"Currently only SQLite, PostgreSQL, MySQL, and FAISS support version-qualified feature references. "
```
Comment (ntkathole) [inline sdk/python/feast/errors.py:145]: probbaly in future PRs we should get rid of this, it needs update everytime

**Review by ntkathole — APPROVED** [review]

File: sdk/python/feast/errors.py (hunk 1 above)
Comment (YassinNouh21) [reply sdk/python/feast/errors.py:145]: Why?

Comment (ntkathole) [reply sdk/python/feast/errors.py:145]: to be less prone to error regarding that message being incorrect or outdated as new support being added
